File: apps/backend/app/models/llm_selection.py
"""
LLM 模型作用域选择响应模型
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

ModelScope = Literal["global", "workspace", "session"]
InheritedScope = Literal["global", "workspace"]


class LLMModelIdentity(BaseModel):
    model_id: Optional[str] = Field(default=None, description="模型配置 ID")
    display_name: Optional[str] = Field(default=None, description="模型显示名称")
    model_name: Optional[str] = Field(default=None, description="底层模型名称")
    provider: Optional[str] = Field(default=None, description="服务商 ID")
    provider_name: Optional[str] = Field(default=None, description="服务商显示名称")


class LLMModelScopeSelection(BaseModel):
    scope: ModelScope
    configured_model_id: Optional[str] = Field(
        default=None,
        description="当前作用域显式保存的模型 ID；为空表示继承上层",
    )
    configured_missing: bool = Field(
        default=False,
        description="显式保存的模型 ID 是否已经失效或不可用",
    )
    configured_display_name: Optional[str] = Field(
        default=None,
        description="显式保存模型的显示名称",
    )
    configured_model_name: Optional[str] = Field(
        default=None,
        description="显式保存模型的底层模型名称",
    )
    configured_provider: Optional[str] = Field(
        default=None,
        description="显式保存模型的服务商 ID",
    )
    configured_provider_name: Optional[str] = Field(
        default=None,
        description="显式保存模型的服务商显示名称",
    )
    inherited_from: Optional[InheritedScope] = Field(
        default=None,
        description="当前作用域未显式指定时，继承自哪一层",
    )
    effective: LLMModelIdentity = Field(
        default_factory=LLMModelIdentity,
        description="按优先级链路解析后的当前生效模型",
    )


class WorkspaceLLMSelectionResponse(BaseModel):
    workspace_id: str
    global_scope: LLMModelScopeSelection
    workspace_scope: LLMModelScopeSelection
    effective: LLMModelIdentity = Field(default_factory=LLMModelIdentity)


class SessionLLMSelectionResponse(BaseModel):
    session_id: str
    workspace_id: Optional[str] = None
    global_scope: LLMModelScopeSelection
    workspace_scope: LLMModelScopeSelection
    session_scope: LLMModelScopeSelection
    effective: LLMModelIdentity = Field(default_factory=LLMModelIdentity)


class UpdateScopedModelSelectionRequest(BaseModel):
    model_id: Optional[str] = Field(
        default=None,
        description="为空时清除当前作用域显式覆盖，回退继承上层",
    )

File: apps/web/src/lib/api/llmSelection.ts
import { API_ENDPOINTS, getCurrentUserId } from "@/config/api";
import { apiRequest } from "@/lib/api/httpClient";

export type LLMSelectionScope = "global" | "workspace" | "session";
export type LLMInheritedScope = "global" | "workspace";

export interface LLMModelIdentitySummary {
  model_id: string | null;
  display_name: string | null;
  model_name: string | null;
  provider: string | null;
  provider_name: string | null;
}

export interface LLMModelScopeSelectionSummary {
  scope: LLMSelectionScope;
  configured_model_id: string | null;
  configured_missing: boolean;
  configured_display_name: string | null;
  configured_model_name: string | null;
  configured_provider: string | null;
  configured_provider_name: string | null;
  inherited_from: LLMInheritedScope | null;
  effective: LLMModelIdentitySummary;
}

export interface WorkspaceLLMSelectionSummary {
  workspace_id: string;
  global_scope: LLMModelScopeSelectionSummary;
  workspace_scope: LLMModelScopeSelectionSummary;
  effective: LLMModelIdentitySummary;
}

export interface SessionLLMSelectionSummary {
  session_id: string;
  workspace_id: string | null;
  global_scope: LLMModelScopeSelectionSummary;
  workspace_scope: LLMModelScopeSelectionSummary;
  session_scope: LLMModelScopeSelectionSummary;
  effective: LLMModelIdentitySummary;
}

export async function getSessionLLMSelection(
  sessionId: string,
  userId = getCurrentUserId(),
): Promise<SessionLLMSelectionSummary> {
  return apiRequest<SessionLLMSelectionSummary>(
    API_ENDPOINTS.SESSION_LLM_SELECTION(userId, sessionId),
    {
      cache: "no-store",
    },
  );
}

export async function updateSessionLLMSelection(
  sessionId: string,
  modelId: string | null,
  userId = getCurrentUserId(),
): Promise<SessionLLMSelectionSummary> {
  return apiRequest<SessionLLMSelectionSummary>(
    API_ENDPOINTS.SESSION_LLM_SELECTION(userId, sessionId),
    {
      method: "PUT",
      body: {
        model_id: modelId,
      },
    },
  );
}

export async function getWorkspaceLLMSelection(
  workspaceId: string,
): Promise<WorkspaceLLMSelectionSummary> {
  return apiRequest<WorkspaceLLMSelectionSummary>(
    API_ENDPOINTS.WORKSPACE_LLM_SELECTION(workspaceId),
    {
      cache: "no-store",
    },
  );
}

export async function updateWorkspaceLLMSelection(
  workspaceId: string,
  modelId: string | null,
): Promise<WorkspaceLLMSelectionSummary> {
  return apiRequest<WorkspaceLLMSelectionSummary>(
    API_ENDPOINTS.WORKSPACE_LLM_SELECTION(workspaceId),
    {
      method: "PUT",
      body: {
        model_id: modelId,
      },
    },
  );
}
